Messagepack wrappers and example executors [Obsolete]

yadudoc opened the pull request:
# Description

## This contents of this PR is in the process of getting stripped into several smaller PRs

This PR is part of a sequence to switch over to using Parsl's HTEX. This PR does the following:

* A new `execution_helper` module with the method: `execute_task(messagepacked_task) -> messagepacked_result`
* Initial version of a wrapped `concurrent.futures.ProcessPoolExecutor`
* Initial version of Parsl's HTEX named `ParslHTEX` that needs to be renamed
* Tests for all of the above

This "shouldn't" break our endpoints yet. Changes to endpoint are pending work on integration tests.


Fixes # (issue)

[sc-21643]
[sc-21994]
[sc-21995]

## Type of change

Choose which options apply, and delete the ones which do not apply.

- New feature (non-breaking change that adds functionality)

yadudoc committed:
Adding bare bones.. edit later

yadudoc committed:
Adding a new helper module with `execute_task` and related tests

yadudoc committed:
Adding a new ProcessPoolExecutor based on concurrent.futures.ProcessPoolExecutor

yadudoc committed:
Adding a new ParslHTEX executor, that uses Parsl's HighThroughputExecutor.

* We are missing functionality to report heartbeats and EPStatusReports

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

khk-globus commented on funcx_endpoint/funcx_endpoint/executors/process_pool/process_pool.py:
```diff
@@ -0,0 +1,109 @@
+        None
+
+        """
+        logger.warning(f"[YADU] task:{future.task_id} completed")
+        self.results_passthrough.put(future.result())
```
Wooo!

yadudoc replied:
I do try to put a personal touch on the warnings :)

khk-globus commented on funcx_endpoint/funcx_endpoint/executors/process_pool/process_pool.py:
```diff
@@ -0,0 +1,109 @@
+        -------
+
+        """
+        unpacked_task: Task = messagepack.unpack(packed_task)
+        future = self.submit(execute_task, {}, unpacked_task.task_id, packed_task)
+
```
As I read the implementation for `self.submit`, won't this add a new item to the `func`'s argument list?  Namely the `unpacked_task.task_id`?

khk-globus reviewed (commented):
This is exciting work!

khk-globus commented on funcx_endpoint/funcx_endpoint/executors/parsl_htex/parsl_htex.py:
```diff
@@ -0,0 +1,50 @@
+        self.results_passthrough.put(future.result())
+
+    def submit_raw(self, packed_task: bytes) -> Future:
+        unpacked_task: Task = messagepack.unpack(packed_task)
+        future = self.submit(execute_task, {}, unpacked_task.task_id, packed_task)
```
Have since been verbally educated that it's _not_ expensive.  So, perhaps this comment is off-base.  On the other hand, if we could engineer it so that the task_id is known to the caller of this method, that "would be nice."

yadudoc replied:
More importantly, we need the task_id at the executor, so that a missing task can be associated with it's task_id.

yadudoc committed:
* execute_task doesn't take task_id arg
* updated mypy errors and tests to match

yadudoc commented on funcx_endpoint/funcx_endpoint/executors/process_pool/process_pool.py:
```diff
@@ -0,0 +1,109 @@
+        return self.executor.shutdown()
+
+
+def double(x):
+    return x * 2
+
+
+p = ProcessPoolExecutor(max_workers=4)
+future = p.submit(double, {}, 5)
```
Removed.

yadudoc commented on funcx_endpoint/funcx_endpoint/executors/process_pool/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from funcx_endpoint.executors.process_pool.process_pool import ProcessPoolExecutor
+
+__all__ = ["ProcessPoolExecutor"]
```
Done

yadudoc commented on funcx_endpoint/funcx_endpoint/executors/execution_helper/helper.py:
```diff
@@ -0,0 +1,107 @@
+
+def call_user_function(
+    message: bytes, result_size_limit: int, serializer=serializer
+) -> str:
```
Done

yadudoc commented on funcx_endpoint/funcx_endpoint/executors/parsl_htex/parsl_htex.py:
```diff
@@ -0,0 +1,50 @@
+
+    def start(self, run_dir: str, results_passthrough: multiprocessing.Queue = None):
+        self.run_dir = os.path.join(os.getcwd(), run_dir)
+        self.run_id = "001"
+        self.provider.script_dir = os.path.join(self.run_dir, "submit_scripts")
+        os.makedirs(self.provider.script_dir, exist_ok=True)
```
There's a bunch of reworked done, but this was because the Parsl's HTEX needed those path to write logs and scripts.

yadudoc commented on funcx_endpoint/funcx_endpoint/executors/parsl_htex/parsl_htex.py:
```diff
@@ -0,0 +1,50 @@
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def start(self, run_dir: str, results_passthrough: multiprocessing.Queue = None):
+        self.run_dir = os.path.join(os.getcwd(), run_dir)
```
Maybe in the next iteration?

yadudoc commented on funcx_endpoint/funcx_endpoint/executors/parsl_htex/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from funcx_endpoint.executors.parsl_htex.parsl_htex import ParslHTEX
+
+__all__ = ["ParslHTEX"]
```
Done.

yadudoc commented on funcx_endpoint/funcx_endpoint/executors/execution_helper/helper.py:
```diff
@@ -0,0 +1,107 @@
+
+def execute_task(
+    task_id: uuid.UUID, task_body: bytes, result_size_limit: int = 10 * 1024 * 1024
+) -> bytes:
```
Probably not. We don't want the user to have control over these limits.

yadudoc commented on funcx_endpoint/funcx_endpoint/executors/execution_helper/helper.py:
```diff
@@ -0,0 +1,107 @@
+
+    f, args, kwargs = serializer.unpack_and_deserialize(task_data)
+    result_data = f(*args, **kwargs)
+    serialized_data = serializer.serialize(result_data)
```
Not sure what would be helpful here. Let's figure this out over a call.

yadudoc commented on funcx_endpoint/funcx_endpoint/executors/execution_helper/helper.py:
```diff
@@ -0,0 +1,107 @@
+
+    else:
+        log.debug("Execution completed without exception")
+        result_message = dict(task_id=task_id, data=result)
```
Fair enough. I've made this change.

yadudoc committed:
Adding a GCExecutorBase class to hold shared functionality:
* Methods to make EPStatusReports
* Threading logic to periodically report EPStatusReports
* `submit_raw` that works as an adapter that allows any executor to execute FuncX Task messages
* callback on future reports results over a passthrough queue

yadudoc committed:
Updating module init files

yadudoc committed:
Minor cleanups only

yadudoc committed:
Renaming `ParslHTEX` -> `GCExecutor`. We agreed to call this an `Engine` on the last call once this is all finalised.
* Added a mock get_status_report which gets reported.

yadudoc committed:
ProcessPoolExecutor now uses the base class

yadudoc committed:
Added tests

yadudoc committed:
Fixing a bunch of mypy errors
* _status_report_thread is of type ReportingThread not threading.Thread
* Making heartbeat_period and optional kwarg on the GCExecutorBase

yadudoc committed:
Removing a log line using the `f"{var=}"` format not supported by py3.7

khk-globus commented on funcx_endpoint/funcx_endpoint/executors/__init__.py:
```diff
@@ -1,3 +1,5 @@
+from funcx_endpoint.executors.gc_executor.gc_executor import GCExecutor
 from funcx_endpoint.executors.high_throughput.executor import HighThroughputExecutor
+from funcx_endpoint.executors.process_pool.process_pool import ProcessPoolExecutor
 
-__all__ = ["HighThroughputExecutor"]
+__all__ = ["HighThroughputExecutor", "ProcessPoolExecutor", "GCExecutor"]
```
`GCExecutor` is almost certainly going to conflict with our rename of `FuncXExecutor`.  Yes, it's an entirely different namespace (`funcx_endpoint.*`), but still perhaps unnecessarily confusing for us devs while in the throes of it, and perhaps slightly odd for administrators to think about in terms of `config.py`, yeah?

Perhaps this is a good time to rename this set of work to `*Engine*`?  Something like:
```
funcx_endpoint/
    engines/
        __init__.py
        globuscompute.py
        highthroughput.py
        processpool.py
```

With an `__all__` in `__init__.py`:
```python
from .globuscompute import GlobusComputeEngine
from .highthroughput import HighThroughputEngine
from .processpool import ProcessPoolEngine

__all__ = (GlobusComputeEngine, HighThroughputEngine, ProcessPoolEngine)
```

yadudoc replied:
We have consensus over on slack #dev that this renaming makes sense. That said, I think I'd like to get this in and have this reorg done in a follow-up PR. Is that okay with you?

khk-globus replied:
As we merged PR #1046 yesterday, I think we can resolve this conversation, yeah?  Feel free to unresolve, if appropriate, and we can discuss further ...

yadudoc commented:
Marking this as no-news since none of these are user-facing changes yet.

yadudoc committed:
Remove bad log line

yadudoc committed:
Executor to engine (#1046)

* No code changes beyond renaming and moving
* funcx_endpoint.executors.GCExecutorBase -> funcx_endpoint.engines.GlobusComputeEngineBase
* funcx_endpoint.executors.ProcessPoolExecutor -> funcx_endpoint.engines.ProcessPoolEngine
* funcx_endpoint.executors.GCExecutor -> funcx_endpoint.engines.GlobusComputeEngine
Similarly the helper script and init files have been moved and aggregated

* Updating tests to use the new engine names.

yadudoc committed:
Skipping gc_engine from engine.submit test due to issue with parsl.htex serialization issue

yadudoc committed:
Add new engines and rework HTEX for backward compatibility (#1057)

* No code changes beyond renaming and moving
* funcx_endpoint.executors.GCExecutorBase -> funcx_endpoint.engines.GlobusComputeEngineBase
* funcx_endpoint.executors.ProcessPoolExecutor -> funcx_endpoint.engines.ProcessPoolEngine
* funcx_endpoint.executors.GCExecutor -> funcx_endpoint.engines.GlobusComputeEngine
Similarly the helper script and init files have been moved and aggregated

* Updating tests to use the new engine names.

* HTEX no longer needs the funcx_service_address with changes from PR #1022

* Update the HTEX.start method to take additional params that are not available at init.

* * Update EPInterchange to require a single executor in config and use that
* `load_config` is removed, since that functionality
 is handled by `start_executor`
* `start_executor` now uses the updated `HTEX.start()` method.

* Stack of minor changes to tests that relied on older behavior.

* Moving over htex from executors to engines

* Mostly renaming
* funcx_endpoint.executors -> funcx_endpoint.engines
* HighThroughputExecutor -> HighThroughputEngine

* Add get_status_report to HTEX

* Updating tests following the executor->engine migration

* * EPInterchange to use queue.Queue for results_passthrough
* EPInterchange to require messages coming via results_passthrough to be messagepacked messages

* * Adding ThreadPoolEngine to module init
* Add ThreadPoolEngine to tests
* Adding a ThreadPoolEngine since the ProcessPoolEngine has issues on MacOS

* Updating HTEX engine to put mpacked Results in results_passthrough and consume mpacked Tasks via submit_raw

* * Adding a bunch of new tests
* Fixing older tests by pointing imports to funcx_endpoint.engines rather than funcx_endpoint.executors
* Some utility remote functions added to tests/utils.py
* Removing bad debug lines
* Minor bug fixes on exception reporting through results_passthrough

* * Minor bugfix. Pass along address kwarg to executor
* Set exception and start exiting thread in ReportingThread
* Aliasing `HighThroughputEngine` as `funcx_endpoint.executors.HighThroughputExecutor` for backward compatibility

* Updating parsl req to 2023.03.27 which has Ben's changes to the signal handler from parsl#2629

* * Ignoring wrong mypy error about address param being duplicated.
